File: packages/python/src/pharos_discovery/adapters/mcp_official.py
from __future__ import annotations

import math
import re
from typing import Any

import httpx

from pharos_discovery.adapters.registry import SearchResult
from pharos_discovery.errors import NoServersFound, RegistryUnavailable
from pharos_discovery.models import (
    AuthSpec,
    Publisher,
    ServerCard,
)

_DEFAULT_REGISTRY = "https://registry.modelcontextprotocol.io/v1"

_TOKEN_RE = re.compile(r"[a-z0-9]+")


def _tokenize(text: str) -> list[str]:
    """Lowercase + extract alphanumeric tokens."""
    return _TOKEN_RE.findall(text.lower())


def _build_corpus_tokens(
    name: str,
    description: str,
    capabilities: list[str],
    tags: list[str],
) -> list[str]:
    """Combine the searchable text fields into a single token list."""
    parts = [name, description, " ".join(capabilities), " ".join(tags)]
    return _tokenize(" ".join(parts))


def _compute_tfidf(
    documents: list[list[str]],
) -> list[dict[str, float]]:
    """Compute TF-IDF vectors for a list of tokenized documents."""
    n_docs = len(documents)
    if n_docs == 0:
        return []

    # Document frequency
    df: dict[str, int] = {}
    for tokens in documents:
        for term in set(tokens):
            df[term] = df.get(term, 0) + 1

    # IDF with smoothing
    idf: dict[str, float] = {
        term: math.log((1 + n_docs) / (1 + count)) + 1.0
        for term, count in df.items()
    }

    vectors: list[dict[str, float]] = []
    for tokens in documents:
        tf: dict[str, int] = {}
        for tok in tokens:
            tf[tok] = tf.get(tok, 0) + 1
        total = len(tokens) if tokens else 1
        vec: dict[str, float] = {}
        for term, count in tf.items():
            vec[term] = (count / total) * idf.get(term, 0.0)
        vectors.append(vec)
    return vectors


def _cosine_similarity(a: dict[str, float], b: dict[str, float]) -> float:
    """Cosine similarity between two sparse TF-IDF vectors."""
    if not a or not b:
        return 0.0
    # Iterate over the smaller vector
    if len(a) > len(b):
        a, b = b, a
    dot = sum(weight * b.get(term, 0.0) for term, weight in a.items())
    norm_a = math.sqrt(sum(w * w for w in a.values()))
    norm_b = math.sqrt(sum(w * w for w in b.values()))
    if norm_a == 0.0 or norm_b == 0.0:
        return 0.0
    return dot / (norm_a * norm_b)


class MCPOfficialAdapter:
    """Adapter for the official MCP registry with client-side semantic re-ranking.

    Fetches server packages from the official MCP registry API
    (https://registry.modelcontextprotocol.io/v1/), normalizes the MCP
    registry format into pharos ``ServerCard`` objects, and applies
    client-side TF-IDF cosine-similarity re-ranking on
    name + description + capabilities + tags.

    This lets agents search the official MCP registry using the same
    pharos discovery interface they use for Pharos-native registries.
    """

    def __init__(
        self,
        base_url: str = _DEFAULT_REGISTRY,
        *,
        search_timeout: float = 10.0,
        get_timeout: float = 10.0,
        api_key: str | None = None,
        client: httpx.AsyncClient | None = None,
    ):
        self._base_url = base_url.rstrip("/")
        self._search_timeout = search_timeout
        self._get_timeout = get_timeout
        self._api_key = api_key
        self._client = client

    @property
    def base_url(self) -> str:
        return self._base_url

    async def search(
        self,
        query: str,
        limit: int = 20,
    ) -> list[SearchResult]:
        """Search the MCP registry and re-rank results by semantic similarity.

        Args:
            query: Free-text search query (matched against name, description,
                capabilities, and tags).
            limit: Maximum number of results to return after re-ranking.

        Returns:
            List of ``SearchResult`` objects sorted by descending
            ``relevanceScore`` (TF-IDF cosine similarity to the query).
            Each result's ``score`` field holds the relevance score.

        Raises:
            NoServersFound: When the registry returns zero packages.
            RegistryUnavailable: On HTTP/network errors.
        """
        raw_items = await self._fetch_packages(query)

        if not raw_items:
            raise NoServersFound(query)

        # Build corpus: one "document" per MCP package
        parsed: list[tuple[ServerCard, list[str]]] = []
        for item in raw_items:
            card = self._normalize(item)
            tokens = _build_corpus_tokens(
                card.display_name,
                card.description,
                card.capabilities,
                card.tags,
            )
            parsed.append((card, tokens))

        # Build the query vector
        query_tokens = _tokenize(query)
        all_docs = [tokens for _, tokens in parsed]
        all_docs.append(query_tokens)
        tfidf_vectors = _compute_tfidf(all_docs)
        query_vec = tfidf_vectors[-1]

        # Score each document
        scored: list[SearchResult] = []
        for idx, (card, _) in enumerate(parsed):
            relevance = _cosine_similarity(query_vec, tfidf_vectors[idx])
            scored.append(SearchResult(card=card, score=relevance))

        # Sort by relevance descending
        scored.sort(key=lambda r: r.score or 0.0, reverse=True)
        return scored[:limit]

    async def get(self, name: str) -> ServerCard:
        """Fetch a single MCP server package by name.

        Args:
            name: The MCP registry package name (e.g. ``@modelcontextprotocol/server-github``).

        Returns:
            A normalized ``ServerCard``.

        Raises:
            RegistryUnavailable: On HTTP errors or when the package is not found.
        """
        headers = self._auth_headers()
        url = f"{self._base_url}/servers/{name}"

        try:
            resp = await self._do_get(url, headers, self._get_timeout)
        except httpx.HTTPError as exc:
            raise RegistryUnavailable(self._base_url, detail=str(exc)) from exc

        if resp.status_code != 200:
            raise RegistryUnavailable(
                self._base_url,
                status=resp.status_code,
                detail=resp.text,
            )

        return self._normalize(resp.json())

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    async def _fetch_packages(self, query: str) -> list[dict[str, Any]]:
        """Fetch raw package list from the MCP registry."""
        headers = self._auth_headers()
        url = f"{self._base_url}/servers"
        params: dict[str, Any] = {}
        if query:
            params["q"] = query

        try:
            resp = await self._do_get(url, headers, self._search_timeout, params=params)
        except httpx.HTTPError as exc:
            raise RegistryUnavailable(self._base_url, detail=str(exc)) from exc

        if resp.status_code != 200:
            raise RegistryUnavailable(
                self._base_url,
                status=resp.status_code,
                detail=resp.text,
            )

        data = resp.json()
        # The MCP registry API returns {"servers": [...]} or {"packages": [...]}
        # Fall back to a bare list for robustness.
        if isinstance(data, list):
            return data
        return data.get("servers", data.get("packages", []))

    async def _do_get(
        self,
        url: str,
        headers: dict[str, str],
        timeout: float,
        params: dict[str, Any] | None = None,
    ) -> httpx.Response:
        if self._client is not None:
            return await self._client.get(url, headers=headers, params=params, timeout=timeout)
        async with httpx.AsyncClient(timeout=timeout) as client:
            return await client.get(url, headers=headers, params=params)

    def _normalize(self, raw: dict[str, Any]) -> ServerCard:
        """Convert an MCP registry package dict into a pharos ``ServerCard``.

        The official MCP registry format differs from the Pharos ServerCard
        schema. This method maps the fields that have equivalents and fills
        sensible defaults for the rest.
        """
        server_id = raw.get("id") or raw.get("name", "unknown")
        name = raw.get("name") or raw.get("display_name") or server_id
        description = raw.get("description") or raw.get("summary") or ""
        version = raw.get("version") or "0.0.0"

        # Publisher
        pub_raw = raw.get("publisher") or {}
        publisher = Publisher(
            id=pub_raw.get("id") or pub_raw.get("name") or "unknown",
            name=pub_raw.get("name") or pub_raw.get("id") or "unknown",
            verified=pub_raw.get("verified"),
            verification_method=pub_raw.get("verification_method"),
            contact=pub_raw.get("contact"),
        )

        # Transport — MCP registry may store as "transports" or "transport"
        raw_transports = raw.get("transports") or raw.get("transport") or []
        if isinstance(raw_transports, str):
            raw_transports = [raw_transports]
        transports = [t for t in raw_transports if t in ("stdio", "http+sse", "streamable-http")]
        if not transports:
            transports = ["http+sse"]

        # Capabilities
        capabilities = raw.get("capabilities") or raw.get("tools") or []

        # Auth
        raw_auth = raw.get("auth") or {}
        auth_type = raw_auth.get("type", "none")
        if auth_type not in ("none", "api_key", "oauth", "mtls"):
            auth_type = "none"
        auth = AuthSpec(type=auth_type)  # type: ignore[arg-type]

        # Status
        raw_status = raw.get("status", "active")
        if raw_status not in ("active", "deprecated", "deleted"):
            raw_status = "active"

        # Endpoint
        endpoint = raw.get("endpoint") or raw.get("url")
        stdio_command = raw.get("stdio_command") or raw.get("command")

        # Timestamps
        published_at = raw.get("published_at") or raw.get("created_at") or ""
        updated_at = raw.get("updated_at") or raw.get("modified_at") or ""
        if not published_at:
            published_at = "1970-01-01T00:00:00Z"
        if not updated_at:
            updated_at = published_at

        # Tags
        tags = raw.get("tags") or []

        # Tools count
        tools_count = raw.get("tools_count", len(capabilities) if capabilities else 0)

        # Source registry
        source_registry = raw.get("source_registry") or self._base_url

        # Availability
        availability = raw.get("availability", "native")
        if availability not in ("mirrored", "referenced", "native"):
            availability = "native"

        return ServerCard(
            id=server_id,
            display_name=name,
            description=description,
            publisher=publisher,
            version=version,
            transport=transports,  # type: ignore[arg-type]
            endpoint=endpoint,
            stdio_command=stdio_command,
            capabilities=capabilities,
            tools_count=tools_count,
            auth=auth,
            availability=availability,  # type: ignore[arg-type]
            tags=tags,
            source_registry=source_registry,
            published_at=published_at,
            updated_at=updated_at,
            status=raw_status,  # type: ignore[arg-type]
        )

    def _auth_headers(self) -> dict[str, str]:
        headers: dict[str, str] = {"Accept": "application/json"}
        if self._api_key:
            headers["Authorization"] = f"Bearer {self._api_key}"
        return headers

File: packages/typescript/src/adapters/mcp-official.ts
import { serverCardSchema, type ServerCard, type AuthSpec, type Publisher } from "../models/index.js";
import { RegistryUnavailable, NoServersFound } from "../errors.js";
import type { SearchResult } from "./registry.js";

export { SearchResult };

const DEFAULT_REGISTRY = "https://registry.modelcontextprotocol.io/v1";

const TOKEN_RE = /[a-z0-9]+/g;

function tokenize(text: string): string[] {
  return text.toLowerCase().match(TOKEN_RE) ?? [];
}

function buildCorpusTokens(
  name: string,
  description: string,
  capabilities: string[],
  tags: string[],
): string[] {
  const combined = [name, description, capabilities.join(" "), tags.join(" ")].join(" ");
  return tokenize(combined);
}

function computeTfidf(documents: string[][]): Map<string, number>[] {
  const nDocs = documents.length;
  if (nDocs === 0) return [];

  // Document frequency
  const df = new Map<string, number>();
  for (const tokens of documents) {
    const unique = new Set(tokens);
    for (const term of unique) {
      df.set(term, (df.get(term) ?? 0) + 1);
    }
  }

  // IDF with smoothing
  const idf = new Map<string, number>();
  for (const [term, count] of df) {
    idf.set(term, Math.log((1 + nDocs) / (1 + count)) + 1.0);
  }

  // TF-IDF vectors
  const vectors: Map<string, number>[] = [];
  for (const tokens of documents) {
    const tf = new Map<string, number>();
    for (const tok of tokens) {
      tf.set(tok, (tf.get(tok) ?? 0) + 1);
    }
    const total = tokens.length || 1;
    const vec = new Map<string, number>();
    for (const [term, count] of tf) {
      vec.set(term, (count / total) * (idf.get(term) ?? 0));
    }
    vectors.push(vec);
  }
  return vectors;
}

function cosineSimilarity(a: Map<string, number>, b: Map<string, number>): number {
  if (a.size === 0 || b.size === 0) return 0.0;
  // Iterate over the smaller map
  const [small, large] = a.size <= b.size ? [a, b] : [b, a];
  let dot = 0.0;
  for (const [term, weight] of small) {
    dot += weight * (large.get(term) ?? 0.0);
  }
  let normA = 0.0;
  for (const w of a.values()) normA += w * w;
  let normB = 0.0;
  for (const w of b.values()) normB += w * w;
  normA = Math.sqrt(normA);
  normB = Math.sqrt(normB);
  if (normA === 0.0 || normB === 0.0) return 0.0;
  return dot / (normA * normB);
}

export interface MCPOfficialAdapterOptions {
  searchTimeout?: number;
  getTimeout?: number;
  apiKey?: string;
}

export class MCPOfficialAdapter {
  private baseUrl: string;
  private searchTimeout: number;
  private getTimeout: number;
  private apiKey: string | null;

  constructor(
    baseUrl: string = DEFAULT_REGISTRY,
    options: MCPOfficialAdapterOptions = {},
  ) {
    this.baseUrl = baseUrl.replace(/\/+$/, "");
    this.searchTimeout = options.searchTimeout ?? 10000;
    this.getTimeout = options.getTimeout ?? 10000;
    this.apiKey = options.apiKey ?? null;
  }

  get url(): string {
    return this.baseUrl;
  }

  async search(query: string, limit: number = 20): Promise<SearchResult[]> {
    const rawItems = await this.fetchPackages(query);

    if (rawItems.length === 0) {
      throw new NoServersFound(query);
    }

    // Build corpus
    const parsed: { card: ServerCard; tokens: string[] }[] = [];
    for (const item of rawItems) {
      const card = this.normalize(item);
      const tokens = buildCorpusTokens(
        card.display_name,
        card.description,
        card.capabilities,
        card.tags,
      );
      parsed.push({ card, tokens });
    }

    // Build query vector
    const queryTokens = tokenize(query);
    const allDocs = parsed.map((p) => p.tokens);
    allDocs.push(queryTokens);
    const tfidfVectors = computeTfidf(allDocs);
    const queryVec = tfidfVectors[tfidfVectors.length - 1];

    // Score each document
    const scored: SearchResult[] = parsed.map((p, idx) => ({
      card: p.card,
      score: cosineSimilarity(queryVec, tfidfVectors[idx]),
    }));

    // Sort by relevance descending
    scored.sort((a, b) => (b.score ?? 0) - (a.score ?? 0));
    return scored.slice(0, limit);
  }

  async get(name: string): Promise<ServerCard> {
    const headers = this.authHeaders();
    const url = `${this.baseUrl}/servers/${encodeURIComponent(name)}`;

    let resp: Response;
    try {
      resp = await fetch(url, {
        method: "GET",
        headers,
        signal: AbortSignal.timeout(this.getTimeout),
      });
    } catch (exc) {
      throw new RegistryUnavailable(this.baseUrl, undefined, String(exc));
    }

    if (!resp.ok) {
      const detail = await resp.text().catch(() => "");
      throw new RegistryUnavailable(this.baseUrl, resp.status, detail);
    }

    const raw = await resp.json();
    return this.normalize(raw);
  }

  // ------------------------------------------------------------------
  // Internal helpers
  // ------------------------------------------------------------------

  private async fetchPackages(query: string): Promise<Record<string, unknown>[]> {
    const headers = this.authHeaders();
    const url = new URL(`${this.baseUrl}/servers`);
    if (query) {
      url.searchParams.set("q", query);
    }

    let resp: Response;
    try {
      resp = await fetch(url.toString(), {
        method: "GET",
        headers,
        signal: AbortSignal.timeout(this.searchTimeout),
      });
    } catch (exc) {
      throw new RegistryUnavailable(this.baseUrl, undefined, String(exc));
    }

    if (!resp.ok) {
      const detail = await resp.text().catch(() => "");
      throw new RegistryUnavailable(this.baseUrl, resp.status, detail);
    }

    const data = await resp.json();
    if (Array.isArray(data)) return data;
    return (data.servers ?? data.packages ?? []) as Record<string, unknown>[];
  }

  private normalize(raw: Record<string, any>): ServerCard {
    const serverId = raw.id ?? raw.name ?? "unknown";
    const name = raw.name ?? raw.display_name ?? serverId;
    const description = raw.description ?? raw.summary ?? "";
    const version = raw.version ?? "0.0.0";

    // Publisher
    const pubRaw = raw.publisher ?? {};
    const publisher: Publisher = {
      id: pubRaw.id ?? pubRaw.name ?? "unknown",
      name: pubRaw.name ?? pubRaw.id ?? "unknown",
      verified: pubRaw.verified ?? undefined,
      verification_method: pubRaw.verification_method ?? undefined,
      contact: pubRaw.contact ?? undefined,
    };

    // Transport
    const rawTransports: unknown[] = raw.transports ?? raw.transport ?? [];
    const validTransports = ["stdio", "http+sse", "streamable-http"] as const;
    let transports = (rawTransports as string[]).filter((t) =>
      validTransports.includes(t as (typeof validTransports)[number]),
    );
    if (transports.length === 0) {
      transports = ["http+sse"];
    }

    // Capabilities
    const capabilities: string[] = raw.capabilities ?? raw.tools ?? [];

    // Auth
    const rawAuth = raw.auth ?? {};
    const validAuthTypes = ["none", "api_key", "oauth", "mtls"] as const;
    const authType = validAuthTypes.includes(rawAuth.type as (typeof validAuthTypes)[number])
      ? rawAuth.type
      : "none";
    const auth: AuthSpec = { type: authType };

    // Status
    const validStatuses = ["active", "deprecated", "deleted"] as const;
    const status = validStatuses.includes(raw.status as (typeof validStatuses)[number])
      ? raw.status
      : "active";

    // Endpoint
    const endpoint = raw.endpoint ?? raw.url ?? undefined;
    const stdioCommand = raw.stdio_command ?? raw.command ?? undefined;

    // Timestamps
    let publishedAt = raw.published_at ?? raw.created_at ?? "";
    let updatedAt = raw.updated_at ?? raw.modified_at ?? "";
    if (!publishedAt) publishedAt = "1970-01-01T00:00:00Z";
    if (!updatedAt) updatedAt = publishedAt;

    // Tags
    const tags: string[] = raw.tags ?? [];

    // Tools count
    const toolsCount: number = raw.tools_count ?? (capabilities.length || 0);

    // Source registry
    const sourceRegistry: string = raw.source_registry ?? this.baseUrl;

    // Availability
    const validAvailabilities = ["mirrored", "referenced", "native"] as const;
    const availability = validAvailabilities.includes(
      raw.availability as (typeof validAvailabilities)[number],
    )
      ? raw.availability
      : "native";

    const cardData = {
      id: serverId,
      display_name: name,
      description,
      publisher,
      version,
      transport: transports,
      endpoint,
      stdio_command: stdioCommand,
      capabilities,
      tools_count: toolsCount,
      auth,
      availability,
      tags,
      source_registry: sourceRegistry,
      published_at: publishedAt,
      updated_at: updatedAt,
      status,
    };

    return serverCardSchema.parse(cardData);
  }

  private authHeaders(): Record<string, string> {
    const headers: Record<string, string> = { Accept: "application/json" };
    if (this.apiKey) {
      headers["Authorization"] = `Bearer ${this.apiKey}`;
    }
    return headers;
  }
}
